### DISCUSSION TITLE
Can we make fishshell a compilable language?

### DISCUSSION THREAD

**Discussion by buddhi-deep** [fish-shell/fish-shell#8867] [Ideas]
Though fish shell is highly depends on linux to run,but i'm still wondering can we make it a compilable language?Though it may need lots of dependences

**Comment by ridiculousfish** [comment]
Sure, [in principle](https://en.wikipedia.org/wiki/Partial_evaluation). What benefit are you thinking of? If it's "distribute a script without a fish dependency" well it would end up as a rather large script. If it's for performance, my traces show that most of fish's time is spent in system calls and memory allocation/deallocation; it doesn't seem limited by its interpreter speed.

**Reply by buddhi-deep** [reply]
Actually what im thinking is to make the script run without fish and it's dependences pre-installed,but still powered/programmed in fishshell

**Reply by ridiculousfish** [reply]
It would be pretty big, a stripped fish binary is ~1.7 MB; if we shipped fish inside the script it would need be at least that large.

**Reply by buddhi-deep** [reply]
Actually, for modern pc this is pretty small, but we can make the script run everywhere

**Reply by faho** [reply]
That "in principle" is understating it quite a bit. This would require making a build of fish that already incorporates a script - but the rest would have to remain intact, it would even have to ship the interactive reader (because of `read`) and a full parser (because of `source` and `eval`). The language is sufficiently dynamic that you can't really remove parts of it except in very special circumstances.

And at that point, you would *still* have a build of fish, that would *still* be linked against a specific ncurses and libc and pcre2. So you couldn't even drop it on a different system (unless it's the same OS, architecture and version). And if you have to compile a specific script for a specific system, you could simply install fish there.

So this would be *a lot* of work that, in practice, is very unlikely to happen (tbh I'm not sure I'd want to merge it even if someone else wrote it and filed a PR) and doesn't even solve the problem. An easier approach is getting fish to work as an appimage (#6475) or similar, so you can move it around as a single file. This would also solve much of the dependency issue.

**Reply by buddhi-deep** [reply]
Oh wow, the appimage way is more likely to be dropped everywhere
Good idea
